pkg/summarizer/summary.go

```diff
@@ -129,67 +129,99 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
 
 	var q config.DashboardQueue
 
-	onConfigUpdate := func(cfg *snapshot.Config) error {
+	fixSnapshot := func(cfg *snapshot.Config) error {
 		dashboards := cfg.AllDashboards()
 		log := log.WithField("onConfigUpdate()", true)
 
 		paths := make([]gcs.Path, 0, len(dashboards))
 		for _, d := range dashboards {
 			path, err := summaryPath(configPath, summaryPathPrefix, d.Name)
 			if err != nil {
-				log.WithError(err).WithField("dashboard", d.Name).Error("bad dashboard path")
+				log.WithError(err).WithField("dashboard", d.Name).Error("Bad dashboard path")
 			}
 			paths = append(paths, *path)
 		}
 
 		stats := gcs.Stat(ctx, client, 10, paths...)
 		whens := make(map[string]time.Time, len(stats))
+		var wg sync.WaitGroup
 		for i, stat := range stats {
 			name := dashboards[i].Name
+			path := paths[i]
+			log := log.WithField("path", path)
 			switch {
 			case stat.Attrs != nil:
 				whens[name] = stat.Attrs.Updated.Add(freq)
 			default:
 				if errors.Is(stat.Err, storage.ErrObjectNotExist) {
-					defer func(i int) {
-						if _, err := lockDashboard(ctx, client, paths[i], 0); err != nil && !gcs.IsPreconditionFailed(err) {
-							log.WithError(err).WithField("path", paths[i]).Error("Failed to lock initial summary")
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						_, err := lockDashboard(ctx, client, path, 0)
+						switch {
+						case gcs.IsPreconditionFailed(err):
+							log.WithError(err).Debug("Lost race to create initial summary")
+						case err != nil:
+							log.WithError(err).Error("Failed to lock initial summary")
+						default:
+							log.Info("Created initial summary")
 						}
-					}(i)
+					}()
 				} else {
-					log.WithError(stat.Err).WithField("path", paths[i]).Info("Failed to stat")
+					log.WithError(stat.Err).Info("Failed to stat")
 				}
 				whens[name] = time.Now()
 			}
 		}
 
+		wg.Wait()
+
 		q.Init(log, cfg.AllDashboards(), time.Now().Add(freq))
 		if err := q.FixAll(whens, false); err != nil {
 			log.WithError(err).Error("Failed to fix all dashboards based on last update time")
 		}
 		return nil
 	}
+	cfgChanged := make(chan *snapshot.Config)
 
+	onConfigUpdate := func(cfg *snapshot.Config) error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case cfgChanged <- cfg:
```

Comment:
I think this observer pattern will work well @chases2!

Reply:
A channel does seem more natural than a callback in golang; a lot fewer race conditions.

Feel free to merge this as-is, but I'm now looking to refactor Observe to use channels instead soon